src/metabase/sfc/analyze.clj

```diff
@@ -0,0 +1,168 @@
+                                  more)))))
+
+(defn field-avg-length
+  "Calculate the average length of all the non-nil strings in a sequence of VALUES."
+  [values]
+  (let [field-values       (filter identity values)
+        field-values-count (count field-values)]
+    (if (zero? field-values-count)
+      0
+      (int (math/round (/ (->> field-values
+                               (map str)
+                               (map count)
+                               (reduce +))
+                          field-values-count))))))
+
+(defn field-fingerprint [driver table field]
+  (let [values      (->> (driver/field-values-lazy-seq driver field)
+                         (take driver/max-sync-lazy-seq-results))
+        field-id    (:id field)
+        fingerprint {:base_type           (:base_type field)
+                     :is_pk               (isa? (:special_type field) :type/PK)
+                     :is_fk               (isa? (:special_type field) :type/FK)
+                     :cardinality         (count (distinct values))
+                     :field_percent_urls  (percent-valid-urls values)
+                     :field_percent_json  (percent-json values)
+                     :field_percent_email (percent-email values)
+                     :field_avg_length    (field-avg-length values)
+                     :field_id            field-id
+                     :table_id            (:id table)
+                     :name                (:name field)
+                     :qualified_name      (field/qualified-name field)
+                     :visibility_type     (:visibility_type field)}]
+    (log/trace (u/format-color 'green "generated fingerprint for field: %s (%s):%s" field-id (:name field) fingerprint))
+    fingerprint))
+
+(defn- save-field-fingerprints!
+  "store a sequence of fingerprints"
+  [fingerprints]
+  (doseq [fingerprint fingerprints]
+    (let [fingerprint (-> fingerprint
+                          (update :base_type u/keyword->qualified-name)
+                          (update :visibility_type u/keyword->qualified-name))]
+         (log/trace (u/format-color 'cyan "saving fingerprint for field: %s (%s):%s"
+                      (:field_id fingerprint) (:name fingerprint) (keys fingerprint)))
+         (or (db/update! FieldFingerprint {:where [:= :field_id (:field_id fingerprint)]
+                                           :set fingerprint})
+             (db/insert! FieldFingerprint fingerprint)))))
+
+(defn- table-fingerprint
+  "generate a fingerprint for a table"
+  [{:keys [rows id name] :as table}]
+  {:rows     rows
+   :table_id id
+   :name     name})
+
+(defn- save-table-fingerprint!
+  "store the table fingerprint for a table
+   field values are stored separately"
+  [fingerprint]
+  (or (db/update! TableFingerprint {:where [:= :table_id (:table_id fingerprint)]
+                                    :set fingerprint})
+      (db/insert! TableFingerprint fingerprint)))
```

Comment:
as mentioned in the other PR this is a potential race condition. Should fix this and the similar code above